File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head: return None
        
        def copy():
            c = head
            while c:
                nex = c.next
                node = Node(c.val)
                c.next = node
                node.next = nex
                c = nex
                
        def random():
            c = head
            while c:
                if c.random:
                    c.next.random = c.random.next
                c = c.next.next
        
        def remove():
            dummy = Node(-1, head.next)
            tail = dummy
            curr = head
            while curr:
                tail.next = curr.next
                tail = tail.next
                curr.next = tail.next
                curr = curr.next
            return dummy.next
        
        copy()
        random()
        return remove()
        

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {

        Map<Node, Node> mp = new HashMap<>();
        mp.put(null, null);

        Node c = head, d = new Node(-1), t = d;
        while (c != null) {
            mp.put(c, new Node(c.val));
            c = c.next;
        }

        c = head;

        while (c != null) {
            t.next = mp.get(c);
            t = t.next;
            t.random = mp.get(c.random);
            c = c.next;
        }
        return d.next;





        
    }
}